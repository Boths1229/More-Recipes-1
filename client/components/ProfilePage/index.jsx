import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Spinner } from 'react-preloading-component';
import lodash from 'lodash';
import ImageUploader from 'react-firebase-file-uploader';
import Tooltip from 'react-tooltip';
import firebase from 'firebase';
import Header from '../common/AuthHeader';
import Footer from '../common/Footer';
import img from '../../public/images/iconic.png';
import verifyUser from '../../util/Authentication';
import {
  getUserRecipes,
  getUserFavRecipes,
  getUserProfile,
  updateUserProfile,
} from '../../actionsCreator/recipes';
import UserDetails from './UserDetails';


const propTypes = {
  user: PropTypes.shape({
    id: PropTypes.number,
    username: PropTypes.string
  }).isRequired,
  actions: PropTypes.shape({
    getUserFavRecipes: PropTypes.func,
    getUserRecipes: PropTypes.func,
    getUserProfile: PropTypes.func,
    updateUserProfile: PropTypes.func
  }).isRequired,
  isLoading: PropTypes.bool.isRequired,
  myRecipesCount: PropTypes.number,
  favRecipesCount: PropTypes.number,
  profile: PropTypes.shape({
    email: PropTypes.string,
    image: PropTypes.string,
    createdAt: PropTypes.string,
  }).isRequired
};

const defaultProps = {
  myRecipesCount: null,
  favRecipesCount: null,
};

/**
 * @description A class to mount the profile page
 * @extends Component
 */
export class ProfilePage extends Component {
  /**
   * @description Creates a recipe
   * @param {object} props
   * @param {object} context
  */
  constructor(props, context) {
    super(props, context);
    this.state = {
      image: '',
      progress: 0,
      isUploading: false,
    };

    this.handleUploadSuccess = this.handleUploadSuccess.bind(this);
    this.handleProgress = this.handleProgress.bind(this);
    this.handleUploadStart = this.handleUploadStart.bind(this);
  }
  /**
   * @description performs an action right after the component mount
   *
   * @memberof ProfilePage
   *
   * @returns {void}
  */
  componentDidMount() {
    const {
      actions
    } = this.props;
    if (verifyUser() === true) {
      actions.getUserRecipes(this.props.user.id, 0);
      actions.getUserFavRecipes(this.props.user.id);
      actions.getUserProfile(this.props.user.id);
    }
  }

  /**
   * @description tracks the progress of uploading image
   *
   * @param {number} progress
   *
   * @memberof ProfilePage
   *
   * @returns {void}
  */
  handleProgress(progress) {
    this.setState({ progress });
  }

  /**
   * @description start the upload operation
   *
   * @memberof ProfilePage
   *
   * @returns {void}
  */
  handleUploadStart() {
    this.setState({ isUploading: true, progress: 0, });
  }

  /**
   * @description get url of image uploaded
   *
   * @param {String} filename
   *
   * @memberof ProfilePage
   *
   * @returns {void}
  */
  handleUploadSuccess(filename) {
    firebase
      .storage()
      .ref('images')
      .child(filename)
      .getDownloadURL()
      .then((url) => {
        this.setState({
          image: url,
          progress: 100,
          isUploading: false,
        });
      })
      .then(() => {
        const data = {
          image: this.state.image
        };
        this.props.actions.updateUserProfile(this.props.user.id, data);
      });
  }

  /**
   * @description renders the components
   *
   * @memberof ProfilePage
   *
   * @returns {JSX} returns the components
  */
  render() {
    const {
      user,
      isLoading,
      profile,
      myRecipesCount,
      favRecipesCount,
    } = this.props;
    return (
      <div>
        <Header user={user} />
        {
          isLoading || lodash.isEmpty(profile)
            ?
              <div style={{
                marginTop: '150px',
                marginLeft: 'auto',
                marginRight: 'auto',
              }}
              >
                <Spinner />
              </div>
            :
              <div className="container recipe-details">
                <section className="row">
                  <div className="col-sm-2" />
                  <div className="col-sm-8">
                    <div id="profile-card" className="card w-100" style={{ width: '20rem' }}>
                      <div className="row pb-3 centered" style={{ marginTop: '10px' }}>
                        <div className="col-sm-12" >
                          <img
                            className="img-fluid user-image rounded-circle"
                            style={{ height: '150px' }}
                            src={profile.image || img}
                            alt="profile"
                          />
                          {
                            this.state.isUploading &&
                            this.state.progress < 100 &&
                            <div>
                              <Spinner />
                            </div>
                          }
                          <div style={{ marginTop: '5px' }}>
                            <label
                              data-tip
                              data-for="upload"
                              className="custom-link"
                            >
                              <i className="fa fa-camera fa-2x vote-button" aria-hidden="true" />
                              <ImageUploader
                                hidden
                                accept="image/*"
                                name="image"
                                storageRef={
                                  firebase
                                    .storage()
                                    .ref('images')
                                }
                                onProgress={this.handleProgress}
                                onUploadSuccess={this.handleUploadSuccess}
                                onUploadStart={this.handleUploadStart}
                              />
                            </label>
                            <Tooltip
                              id="upload"
                              place="top"
                              type="dark"
                            >
                              <span>Upload image</span>
                            </Tooltip >
                          </div>
                        </div>
                      </div>
                      <hr />
                      <UserDetails
                        username={user.username}
                        email={profile.email}
                        myRecipesCount={myRecipesCount}
                        favoriteRecipeCount={favRecipesCount}
                        createdAt={profile.createdAt.split('T')[0]}
                      />
                      <hr />
                    </div>
                  </div>
                </section>
              </div>
        }

        <Footer />
      </div>
    );
  }
}

/**
 * @description maps state to properties of profilePage
 *
 * @param  {object} state
 *
 * @returns {object} returns the state to be mapped to props
 */
function mapStateToProps(state) {
  return {
    user: state.auth.user,
    favRecipesCount: state.recipes.favoriteRecipeCount,
    myRecipesCount: state.recipes.userRecipesCount,
    profile: state.auth.profile,
    isLoading: state.auth.isLoading
  };
}

/**
 * @description maps action to properties of MyFavoriteRecipePage
 *
 * @param  {object} dispatch
 *
 * @returns {object} returns the action to be bind
 */
function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      getUserRecipes,
      getUserFavRecipes,
      getUserProfile,
      updateUserProfile
    }, dispatch)
  };
}

ProfilePage.propTypes = propTypes;
ProfilePage.defaultProps = defaultProps;


export default connect(mapStateToProps, mapDispatchToProps)(ProfilePage);
