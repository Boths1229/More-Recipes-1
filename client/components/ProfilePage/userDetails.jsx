import React, { PropTypes } from 'react';

const propTypes = {
  username: PropTypes.string.isRequired,
  email: PropTypes.string.isRequired,
  myRecipesCount: PropTypes.number.isRequired,
  favRecipesCount: PropTypes.number.isRequired,
  createdAt: PropTypes.string.isRequired,
};

const userDetails = props => (
  <div className="row">
    <div className="col-sm-12">
      <div className="row">
        <div className="col-sm-12">
          <h5>
            username: <span>{props.username}</span>
          </h5>
        </div>
        <div className="col-sm-12">
          <h5>
            Email: <span>{props.email}</span>
          </h5>
        </div>
        <div className="col-sm-12">
          <h5>
            Recipe Added: <span>{props.myRecipesCount}</span>
          </h5>
        </div>
        <div className="col-sm-12">
          <h5>
            Favorite Recipes: <span>{props.favRecipesCount}</span>
          </h5>
        </div>
        <div className="col-sm-12">
          <h5>
            Date Created: <span>{props.createdAt}</span>
          </h5>
        </div>
      </div>
    </div>
  </div>
);

userDetails.propTypes = propTypes;
export default userDetails;
