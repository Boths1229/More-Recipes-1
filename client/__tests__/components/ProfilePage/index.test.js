import React from 'react';
import { shallow } from 'enzyme';
import { Spinner } from 'react-preloading-component';
import { ProfilePage } from '../../../components/ProfilePage/index';
import MockData from '../../__mocks__/actions/recipes';


let props;

const setup = () => {
  props = {
    user: MockData.userData.profile,
    actions: {
      getUserFavRecipes: () => {},
      getUserRecipes: () => {},
      getUserProfile: () => {},
      updateUserProfile: () => {}
    },
    myRecipesCount: 3,
    favRecipesCount: 4,
    createdAt: '2018-02-20T10:11:25.020Z',
    isLoading: false,
    profile: MockData.userData.profile
  };

  return shallow(<ProfilePage {...props} />);
};


localStorage.clear = () => {};
localStorage.getItem = () => {};
localStorage.setItem = () => {};

describe('ProfilePage Component', () => {
  it('should render correctly', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should set state of isUploading and progress when image upload starts', () => {
    const wrapper = setup();
    const action = wrapper.instance();
    const startUpload = jest.spyOn(action, 'handleUploadStart');
    action.handleUploadStart();
    expect(startUpload).toBeCalled();
    expect(wrapper.state('isUploading')).toBe(true);
    expect(wrapper.state('progress')).toBe(0);
  });

  it('should track state of progress when image is uploading', () => {
    const wrapper = setup();
    const action = wrapper.instance();
    const trackProgress = jest.spyOn(action, 'handleProgress');
    action.handleProgress(20);
    expect(trackProgress).toBeCalled();
    expect(wrapper.state('progress')).toBe(20);
  });

  it('should render loader when `isLoading` is true', () => {
    const wrapper = setup();
    wrapper.setProps({ isLoading: true });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Spinner).length).toBe(1);
  });
});
