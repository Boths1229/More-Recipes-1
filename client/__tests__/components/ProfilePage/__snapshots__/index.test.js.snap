// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePage Component should render correctly 1`] = `
<div>
  <Connect(AuthHeader)
    user={
      Object {
        "createdAt": "2018-02-19T15:54:05.322Z",
        "email": "tester@test.com",
        "id": 1,
        "image": null,
        "username": "sannytee",
      }
    }
  />
  <div
    className="container recipe-details"
  >
    <section
      className="row"
    >
      <div
        className="col-sm-2"
      />
      <div
        className="col-sm-8"
      >
        <div
          className="card w-100"
          id="profile-card"
          style={
            Object {
              "width": "20rem",
            }
          }
        >
          <div
            className="row pb-3 centered"
            style={
              Object {
                "marginTop": "10px",
              }
            }
          >
            <div
              className="col-sm-12"
            >
              <img
                alt="profile"
                className="img-fluid user-image rounded-circle"
                src="test"
                style={
                  Object {
                    "height": "150px",
                  }
                }
              />
              <div
                style={
                  Object {
                    "marginTop": "5px",
                  }
                }
              >
                <label
                  className="custom-link"
                  data-for="upload"
                  data-tip={true}
                >
                  <i
                    aria-hidden="true"
                    className="fa fa-camera fa-2x vote-button"
                  />
                  <FirebaseFileUploader
                    accept="image/*"
                    hidden={true}
                    name="image"
                    onProgress={[Function]}
                    onUploadStart={[Function]}
                    onUploadSuccess={[Function]}
                    storageRef={
                      Object {
                        "ref": [MockFunction] {
                          "calls": Array [
                            Array [
                              "images",
                            ],
                          ],
                        },
                        "storage": [MockFunction] {
                          "calls": Array [
                            Array [],
                          ],
                        },
                      }
                    }
                  />
                </label>
                <ReactTooltip
                  id="upload"
                  insecure={true}
                  place="top"
                  resizeHide={true}
                  type="dark"
                  wrapper="div"
                >
                  <span>
                    Upload image
                  </span>
                </ReactTooltip>
              </div>
            </div>
          </div>
          <hr />
          <userDetails
            createdAt="2018-02-19"
            email="tester@test.com"
            favoriteRecipeCount={4}
            myRecipesCount={3}
            username="sannytee"
          />
          <hr />
        </div>
      </div>
    </section>
  </div>
  <Footer />
</div>
`;

exports[`ProfilePage Component should render loader when \`isLoading\` is true 1`] = `
<div>
  <Connect(AuthHeader)
    user={
      Object {
        "createdAt": "2018-02-19T15:54:05.322Z",
        "email": "tester@test.com",
        "id": 1,
        "image": null,
        "username": "sannytee",
      }
    }
  />
  <div
    style={
      Object {
        "marginLeft": "auto",
        "marginRight": "auto",
        "marginTop": "150px",
      }
    }
  >
    <n
      color="#FF4601"
      secondaryColor="#e6e6e6"
      size={40}
    />
  </div>
  <Footer />
</div>
`;
